src/System.Security.Cryptography.OpenSsl/src/System/Security/Cryptography/ECDsaOpenSsl.cs

```diff
@@ -265,107 +292,110 @@ private static void CopySignatureField(byte[] signatureField, byte[] response, i
             }
         }
 
-        private void FreeKey()
+        public override int KeySize
         {
-            if (_key != null && _key.IsValueCreated)
+            get
             {
-                SafeEcKeyHandle handle = _key.Value;
+                return base.KeySize;
+            }
+            set
+            {
+                if (KeySize == value)
+                    return;
 
-                if (handle != null)
-                {
-                    handle.Dispose();
-                }
+                // Set the KeySize before FreeKey so that an invalid value doesn't throw away the key
+                base.KeySize = value;
+
+                FreeKey();
+                _key = new Lazy<SafeEcKeyHandle>(GenerateKeyLazy);
             }
         }
 
-        private static int GetKeySize(SafeEcKeyHandle ecKeyHandle)
+        public override void GenerateKey(ECCurve curve)
         {
-            int nid = Interop.Crypto.EcKeyGetCurveName(ecKeyHandle);
-            int keySize = 0;
+            curve.Validate();
+            FreeKey();
```

Comment:
Okay, SetKey doesn't currently FreeKey, so "we should move FreeKey to SetKey" :)